### PR TITLE
Jack/discover2

### PR DESCRIPTION
**Pull request by jackthmp**
- Implements discover designs
- Adds search bar, podcast cells, search results view
- Adds KingFisher pod for downloading images

### CODE REVIEW

File: Recast/Discover/DiscoverTableViewHeader.swift [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        podcastTitleLabel.textColor = .white
+        episodeDescriptionView.addSubview(podcastTitleLabel)
+
+        episodeTitleLabel = UILabel()
+        episodeTitleLabel.font = .systemFont(ofSize: 16)
+        episodeTitleLabel.textColor = .gray
+        episodeDescriptionView.addSubview(episodeTitleLabel)
+
+        subscribeButton = UIButton()
+        subscribeButton.backgroundColor = #colorLiteral(red: 0.2823529412, green: 0.6117647059, blue: 0.5960784314, alpha: 1)
```
Comment (mindylou) [inline Recast/Discover/DiscoverTableViewHeader.swift]: change this to zeplin's color later (I think @jaewonsim made a constant for it)

File: Recast/Discover/DiscoverTableViewHeader.swift [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        episodeDescriptionView.addSubview(subscribeButton)
+
+        episodeDescriptionLabel = UILabel()
+        episodeDescriptionLabel.numberOfLines = 3
+        episodeDescriptionLabel.font = .systemFont(ofSize: 16)
+        episodeDescriptionLabel.textColor = .white
+        episodeDescriptionView.addSubview(episodeDescriptionLabel)
+
+        seeMoreButton = UIButton()
+        seeMoreButton.setTitle("see more", for: .normal)
```
Comment (mindylou) [inline Recast/Discover/DiscoverTableViewHeader.swift]: change this to "See more"

File: Recast/Discover/DiscoverTableViewHeader.swift [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        seeMoreButton.snp.makeConstraints { make in
+            make.trailing.equalTo(episodeDescriptionLabel.snp.trailing)
+            make.top.equalTo(episodeDescriptionLabel.snp.bottom).offset(episodeTitleTopPadding)
+        }
+    }
+
+    @objc func expandDescription() {
+        isDescriptionExpanded.toggle()
+        self.episodeDescriptionLabel.numberOfLines = isDescriptionExpanded ? 0 : 3
+
```
Comment (mindylou) [inline Recast/Discover/DiscoverTableViewHeader.swift]: delete newline

File: Recast/Discover/DiscoverViewController.swift [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+        tableViewHeader = DiscoverTableViewHeader(frame: .zero)
+        view.addSubview(tableViewHeader)
+
+        makeConstraints()
+    }
+
+    func makeConstraints() {
+        tableViewHeader.snp.makeConstraints { make in
+            //it was all in the console. It told me the height was ambigious
```
Comment (mindylou) [inline Recast/Discover/DiscoverViewController.swift]: you can delete this comment from austin lol

File: Recast/Discover/MainSearchDataSourceDelegate.swift [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+class MainSearchDataSourceDelegate: NSObject {
+
+    // MARK: - Variables
+    weak var delegate: SearchTableViewDelegate?
+    var searchResults: [PartialPodcast] = []
+
+    func fetchData(query: String) {
+        SearchEndpoint(parameters: ["term": query, "media": "podcast", "limit": -1]).run()
+            .success { (response: SearchResults) in
```
Comment (mindylou) [inline Recast/Discover/MainSearchDataSourceDelegate.swift]: you can just do `response in`

File: Recast/Discover/MainSearchDataSourceDelegate.swift [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                self.searchResults = response.results
+                self.delegate?.refreshController()
+            }
+            .failure { (error: Error) in
```
Comment (mindylou) [inline Recast/Discover/MainSearchDataSourceDelegate.swift]: `error in`

File: Recast/Discover/MainSearchViewController.swift [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+    func makeConstraints() {
+
+        searchResultsTableView.snp.makeConstraints { make in
+            make.edges.equalToSuperview()
+        }
+        discoverContainerView.snp.makeConstraints { make in
+            make.edges.equalToSuperview()
+        }
+
```
Comment (mindylou) [inline Recast/Discover/MainSearchViewController.swift]: delete newline

File: Recast/Discover/MainSearchViewController.swift [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        discoverVC = DiscoverViewController()
+        addChildViewController(discoverVC)
+        discoverContainerView.addSubview(discoverVC.view)
+
+        makeConstraints()
+    }
+
+    func makeConstraints() {
+
```
Comment (mindylou) [inline Recast/Discover/MainSearchViewController.swift]: delete newline

File: Recast/Discover/MainSearchDataSourceDelegate.swift [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+        }
+    }
+
+    func resetResults() {
+        searchResults = []
+    }
+
+}
+
+extension MainSearchDataSourceDelegate: UITableViewDataSource {
```
Comment (mindylou) [inline Recast/Discover/MainSearchDataSourceDelegate.swift:38]: Add `// MARK:` for these

**Review by mindylou — COMMENTED** [review]
Small changes but otherwise looks good

File: Recast/AppDelegate.swift [hunk 10]
```diff
@@ -13,18 +13,22 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+
+        if let statusBar = UIApplication.shared.value(forKeyPath: "statusBarWindow.statusBar")
+            as? UIView {
+            statusBar.backgroundColor = .clear
+        }
+
         window?.rootViewController = navController
         window?.makeKeyAndVisible()
-        
+
```
Comment (mindylou) [inline Recast/AppDelegate.swift:30]: make sure to run `swiftlint autocorrect`!

**Review by mindylou — DISMISSED** [review]
Remember to run swiftlint autocorrect but otherwise lgtm

File: Recast/Discover/MainSearchDataSourceDelegate.swift [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+                print(error)
+        }
+    }
+
+    func resetResults() {
+        searchResults = []
+    }
+
+}
+
```
Comment (mindylou) [inline Recast/Discover/MainSearchDataSourceDelegate.swift]: Also add a `// MARK` for here

File: PreviewPlayerView.swift [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+//
+
+import UIKit
+import SnapKit
+
+class PreviewPlayerView: UIView {
+
+    // MARK: - Variables
+    var upNextView: UIView!
+    private var upNextLabel: UILabel!
```
Comment (doctor3w) [inline PreviewPlayerView.swift]: No need to make these private, let's leave these public until we know for sure how it'll interact with our player and view controller.

File: Recast/Cells/PodcastTableViewCell.swift [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+        setUpConstraints()
+    }
+
+    func setUpConstraints() {
+        // MARK: - Constants
+        let padding = 10
+        let imageHeight = 75
+
+        super.layoutSubviews()
```
Comment (doctor3w) [inline Recast/Cells/PodcastTableViewCell.swift]: don't need this do we?

File: Recast/Discover/MainSearchDataSourceDelegate.swift [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+// MARK: - UITableViewDataSource
+extension MainSearchDataSourceDelegate: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return searchResults.count
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        //swiftlint:disable:next line_length
+        let cell = tableView.dequeueReusableCell(withIdentifier: PodcastTableViewCell.cellReuseIdentifier, for: indexPath)
```
Comment (doctor3w) [inline Recast/Discover/MainSearchDataSourceDelegate.swift]: You should be able to just force cast this, shouldn't have to safely fall to a blank cell

File: Recast/Discover/MainSearchDataSourceDelegate.swift [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+            as? PodcastTableViewCell ?? PodcastTableViewCell()
+        let podcast = searchResults[indexPath.row]
+
+        // Set up cell
+        let artworkURL = podcast.artworkUrl600 ?? podcast.artworkUrl100 ?? podcast.artworkUrl60 ?? podcast.artworkUrl30
```
Comment (doctor3w) [inline Recast/Discover/MainSearchDataSourceDelegate.swift:50]: dat chaining tho!

File: Recast/Discover/MainSearchViewController.swift [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+        searchField?.textColor = .white
+        searchField?.backgroundColor = #colorLiteral(red: 0.09749762056, green: 0.09749762056, blue: 0.09749762056, alpha: 1)
+
+        navigationItem.titleView = searchController?.searchBar
+
+        searchResultsTableView = UITableView(frame: .zero, style: .plain)
+        searchResultsTableView.dataSource = tableViewData
+        searchResultsTableView.delegate = tableViewData
+        //swiftlint:disable:next line_length
+        searchResultsTableView.register(PodcastTableViewCell.self, forCellReuseIdentifier: PodcastTableViewCell.cellReuseIdentifier)
```
Comment (doctor3w) [inline Recast/Discover/MainSearchViewController.swift]: You could probably decrease this line length by pulling PodcastTableViewCell.cellReuseIdentifier out into a `let = reuseId = ...`

Better than ignoring the convention

File: Recast/PlayerControlsView.swift [hunk 17]
```diff
@@ -83,6 +83,7 @@ class PlayerControlsView: UIView {
         fatalError("init(coder:) has not been implemented")
     }
 
+    //swiftlint:disable:next function_body_length
```
Comment (doctor3w) [inline Recast/PlayerControlsView.swift]: space after `//`

File: Recast/SeriesViewController.swift [hunk 18]
```diff
@@ -64,15 +64,15 @@ extension SeriesViewController: UITableViewDataSource {
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
         return EpisodeTableViewCell(style: .default, reuseIdentifier: episodeCellReuseIdentifer)
```
Comment (doctor3w) [inline Recast/SeriesViewController.swift]: dequeue?

File: Recast/iTunesPodcasts/iTunesSearch/Models.swift [hunk 19]
```diff
@@ -60,7 +60,7 @@ struct SearchResult: Codable, PartialPodcast {
     let country: String
     let currency: String
     let primaryGenreName: String!
-    let contentAdvisoryRating: String
+    let contentAdvisoryRating: String?
```
Comment (doctor3w) [inline Recast/iTunesPodcasts/iTunesSearch/Models.swift:63]: Is this not always there? Interesting...

**Review by doctor3w — CHANGES_REQUESTED** [review]

**Review by doctor3w — APPROVED** [review]

**Review by doctor3w — APPROVED** [review]

**Pull request merged**